pkg/session_recording/player.jsx

```diff
@@ -0,0 +1,823 @@
+             * receives a packet at the specified index, the matching tuple is
+             * removed from the list, and its deferred object is resolved.
+             * This is used to keep users informed about packets arriving.
+             */
+            this.idxDfdList = [];
+            /* Last seen message ID */
+            this.id = 0;
+            /* Last seen time position */
+            this.pos = 0;
+            /* Last seen window width */
+            this.width = null;
+            /* Last seen window height */
+            this.height = null;
+            /* List of packets read */
+            this.pktList = [];
+            /* Error which stopped the loading */
+            this.error = null;
+            /* The journalctl reading the recording */
+            this.journalctl = Journal.journalctl(
+                                this.matchList,
+                                {count: "all", follow: false});
+            this.journalctl.fail(this.handleError);
+            this.journalctl.stream(this.handleStream);
+            this.journalctl.done(this.handleDone);
+            /*
+             * Last seen cursor of the first, non-follow, journalctl run.
+             * Null if no entry was received yet, or the second run has
+             * skipped the entry received last by the first run.
+             */
+            this.cursor = null;
+            /* True if the first, non-follow, journalctl run has completed */
+            this.done = false;
+        }
+
+        /*
+         * Return a promise which is resolved when a packet at a particular
+         * index is received by the buffer. The promise is rejected with a
+         * non-null argument if an error occurs or has occurred previously.
+         * The promise is rejected with null, when the buffer is stopped. If
+         * the packet index is not specified, assume it's the next packet.
+         */
+        awaitPacket(idx) {
+            let i;
+            let idxDfd;
+
+            /* If an error has occurred previously */
+            if (this.error !== null) {
+                /* Reject immediately */
+                return $.Deferred().reject(this.error).promise();
+            }
+
+            /* If the buffer was stopped */
+            if (this.journalctl === null) {
+                return $.Deferred().reject(null).promise();
+            }
+
+            /* If packet index is not specified */
+            if (idx === undefined) {
+                /* Assume it's the next one */
+                idx = this.pktList.length;
+            } else {
+                /* If it has already been received */
+                if (idx < this.pktList.length) {
+                    /* Return resolved promise */
+                    return $.Deferred().resolve().promise();
+                }
+            }
+
+            /* Try to find an existing, matching tuple */
+            for (i = 0; i < this.idxDfdList.length; i++) {
```

Comment:
Tuple here consists from different packets, right?

Reply:
A tuple here consists of a packet index and a Deferred object. I described `this.idxDfdList` in the constructor. Would it be better to put the description somewhere else, perhaps?

Reply:
Let's leave it as is.